src/inmanta/server/services/environmentservice.py

```diff
@@ -194,6 +199,29 @@ async def delete_environment(self, environment_id: uuid.UUID) -> Apireturn:
         await self.environment_delete(environment_id)
         return 200
 
+    @protocol.handle(methods_v2.halt_environment, env="tid")
+    async def halt(self, env: data.Environment) -> None:
+        async with self.agent_state_lock:
+            if env.halted:
+                return
+            async with data.Environment.get_connection() as connection:
+                async with connection.transaction():
+                    await env.update_fields(halted=True, connection=connection)
+                    await self.agent_manager.halt_agents(env, connection=connection)
+        await self.autostarted_agent_manager.stop_agents(env)
+
+    @protocol.handle(methods_v2.resume_environment, env="tid")
+    async def resume(self, env: data.Environment) -> None:
+        async with self.agent_state_lock:
+            if not env.halted:
+                return
+            async with data.Environment.get_connection() as connection:
+                async with connection.transaction():
+                    await env.update_fields(halted=False, connection=connection)
+                    await self.agent_manager.resume_agents(env, connection=connection)
+        await self.autostarted_agent_manager.restart_agents(env)
+        await self.server_slice.compiler.resume_environment(env.id)
```

Comment:
Same here.

Reply:
In this case I think we only need to add `restart_agents` under the lock. `compiler.resume_environment` has nothing to do with agent state and the compiler service has it's own locks.

Reply:
Agree.